src/combinatorics/catalan.lean

```diff
@@ -0,0 +1,158 @@
+begin rw catalan', simp, end
+
+/-- A helper sequence that can be used to prove the equality of the recursive and the explicit
+definition using a telescoping sum argument. -/
+private def gosper_catalan (n j : ℕ) : ℚ :=
+nat.central_binom j * nat.central_binom (n - j) * (2 * j - n) / (2 * n * (n + 1))
+
+lemma gosper_trick (n i : ℕ) (h : i ≤ n) :
+gosper_catalan (n+1) (i+1) - gosper_catalan (n+1) i = catalan' i * catalan' (n - i) :=
+begin
+rw ← @mul_right_inj' _ _ (i + 1 : ℚ) _ _ (nat.cast_add_one_ne_zero i),
+rw ← @mul_right_inj' _ _ (n - i + 1 : ℚ) _ _ (by exact_mod_cast (n - i).succ_ne_zero),
+{ simp only [gosper_catalan, catalan'],
+  rw [mul_sub, mul_sub],
+  rw mul_div,
+  rw ← mul_assoc,
+  rw ← mul_assoc,
+  rw_mod_cast nat.succ_mul_central_binom_succ i,
+  rw mul_comm i.central_binom,
+  simp only [← mul_assoc],
+  rw_mod_cast mul_comm (n - i + 1),
+  nth_rewrite 1 mul_div,
+  push_cast,
+  simp only [← mul_assoc],
+  rw mul_assoc (i + 1 : ℚ),
+  simp_rw nat.sub_add_comm h,
+  rw_mod_cast nat.succ_mul_central_binom_succ,
+  simp_rw nat.add_sub_add_right,
+  rw mul_comm _ (n-i).central_binom,
+  rw mul_comm _ (n-i).central_binom,
+  rw ← mul_assoc (i+1),
+  rw mul_comm (i+1),
+  push_cast,
+  simp only [mul_div, ← mul_assoc],
+  rw_mod_cast mul_comm _ (n-i).central_binom,
+  simp only [← mul_div, ← mul_assoc],
+  push_cast,
+  simp only [mul_assoc],
+  rw ← mul_sub,
+  simp only [← mul_assoc],
+  rw mul_comm _ ((n-i).central_binom : ℚ),
+  congr,
+  rw mul_comm _ (i.central_binom : ℚ),
+  rw mul_comm _ (i.central_binom : ℚ),
+  rw mul_comm _ (i.central_binom : ℚ),
+  simp only [mul_assoc],
+  rw ← mul_sub,
+  congr,
+  ring_nf,
+  field_simp,
+  rw div_self,
+  { have h : ((n : ℚ) + (-(i : ℚ) + 1)) * (i + 1) = (i + 1) * n + (-i ^ 2 + 1),
+    { ring },
+    { rw [← h, div_self],
+      apply mul_ne_zero,
+      { rw ← add_assoc,
+        rw ← sub_eq_add_neg,
+        exact_mod_cast (n-i).succ_ne_zero, },
+      { exact_mod_cast i.succ_ne_zero } }, },
+  { exact_mod_cast nat.succ_ne_zero _ } },
+end
+
+lemma gosper_catalan_sub_eq_catalan' (n : ℕ) :
+  gosper_catalan (n + 1) (n + 1) - gosper_catalan (n + 1) 0 = catalan' (n + 1) :=
+begin
+simp only [gosper_catalan, catalan', nat.cast_succ, one_div, tsub_self, nat.central_binom_zero,
+  nat.cast_one, mul_one, tsub_zero, one_mul, nat.cast_zero, mul_zero, zero_sub, neg_add_rev,
+  div_eq_mul_inv],
+rw [mul_assoc, mul_assoc, mul_assoc, ← mul_sub, mul_comm],
+congr,
+nth_rewrite 1 ← one_mul (n + 1 : ℚ),
+rw [← neg_add_rev, neg_mul, sub_neg_eq_add, ← sub_mul],
+norm_num,
+rw [← mul_add, ← add_mul, ← two_mul, mul_comm _ (1/2 : ℚ), ← mul_assoc, mul_comm (n + 1 + 1 : ℚ)⁻¹],
+simp only [← mul_assoc],
+nth_rewrite 1 ← one_mul (n + 1 + 1 : ℚ)⁻¹,
+congr,
+simp only [one_div, inv_mul_cancel_right₀, ne.def, bit0_eq_zero, one_ne_zero, not_false_iff],
+apply mul_inv_cancel,
+exact_mod_cast n.succ_ne_zero,
+end
+
+theorem catalan_defs_agree (n : ℕ) :
+catalan' n = catalan n :=
```

Comment:
Ultimately I don't think we want both `catalan` and `catalan`, and that this lemma should be written as
```suggestion
theorem catalan_eq_central_binom_div (n : ℕ) :
  catalan' n = nat.central_binom n / (n + 1) :=
```

Reply:
Why not both? Prove the divisibility in that one file with all the lemmas about binomials and divisibility, then prove `catalan n = n.central_binom / (n + 1)`.

I agree that it's kind of dumb to have two definitions for the same thing, and since `catalan` gives an induction principle, that one should stay.